keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1107 @@
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This proposes a component leader election mechanism that is safer for upgrades
+and rollbacks.
+
+This leader election approach continues to use leases, but with two
+key modifications:
+
+- Instead of a race by component instances to claim the lease, component instances
+  declare candidacy for a lease and a election coordinator claims the lease for the
+  best available candidate. This allows the election coordinator to pick a
+  candidate with the lowest version to ensure that skew rules are not violated.
```

Comment:
how can we determine component versions today without looking at `ps` or configuration files?
for example, KCM and kube-scheduler.

Reply:
it looks like candidates have to be more aware of themselves. I think the election coordination requires a separate object type that is flexible enough to add more rules, evaluated based on the information sent from candidates.

Reply:
> how can we determine component versions today without looking at ps or configuration files?

For version awareness of the KCM and kube-scheduler we can leverage https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/version/version.go

Reply:
> I think the election coordination requires a separate object type that is flexible enough to add more rules, evaluated based on the information sent from candidates.

There a trade off here.  If we continue to use Lease, we have to put any additional information we need into metadata.annotations.  Leases already do quite a bit of this.  API server identity is a good example of where we've gone this route.   If we introduce a new object type, we could define the exact fields we need, but we'd diverge from how other parts of the system use leases, and we'd have a more complex migration path... the migration path if we continue to use the same leases is relatively trivial.

Reply:
> For version awareness of the KCM and kube-scheduler we can leverage https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/version/version.go

as in, calling `component-name --version` on the host, to determine the version is running there?

Reply:
> as in, calling component-name --version on the host, to determine the version that is running there?

Components publish version information into [identity leases](https://github.com/kubernetes/enhancements/blob/529c9d675c5e2214a9320f80a1dbbdcb77f1a288/keps/sig-api-machinery/4355-coordinated-leader-election/README.md#component-identity-leases), the election controller observes the identity leases when managing the leader lease.

Reply:
Do we have to restrict the coordinated leader election mechanism to only understand version as the only criteria to decide leader? Would it be beneficial to add a rule structure so that cluster operators can extend the rules beyond version ?

Reply:
> Would it be beneficial to add a rule structure so that cluster operators can extend the rules beyond version ?

Maybe?  What sort of use cases do you have in mind?